src/Tokenizer.php

```diff
@@ -726,57 +727,108 @@ final class Tokenizer
+        self::$nextTokenRegexReservedToplevel  = '/\G' . $regexReservedToplevel . '(?=$|\s|' . $regexBoundaries . ')/';
+        self::$nextTokenRegexReservedNewline   = '/\G' . $regexReservedNewline . '(?=$|\s|' . $regexBoundaries . ')/';
+        self::$nextTokenRegexReserved          = '/\G' . $regexReserved . '(?=$|\s|' . $regexBoundaries . ')/';
+        self::$nextTokenRegexFunction          = '/\G' . $regexFunction . '(?=\s*\()/';
+        self::$nextTokenRegexNonReserved       = '/\G.*?(?=$|\s|["\'`]|' . $regexBoundaries . ')/';
+    }
+
+    /** @param list<string> $values */
+    private function makeRegexFromList(array $values, bool $sorted = false): string
+    {
+        // sort list by alphabet and from longest word to shortest
+        if (! $sorted) {
+            usort($values, static function (string $a, string $b) {
+                return str_starts_with($a, $b) || str_starts_with($b, $a)
+                    ? strlen($b) <=> strlen($a)
+                    : $a <=> $b;
+            });
+        }
+
+        /** @var array<int|string, list<string>> $valuesBySharedPrefix */
+        $valuesBySharedPrefix = [];
+        $items                = [];
+        $prefix               = null;
+
+        foreach ($values as $v) {
+            if ($prefix !== null && ! str_starts_with($v, substr($prefix, 0, 1))) {
+                $valuesBySharedPrefix[$prefix] = $items;
+                $items                         = [];
+                $prefix                        = null;
+            }
+
+            $items[] = $v;
+
+            if ($prefix === null) {
+                $prefix = $v;
+            } else {
+                while (! str_starts_with($v, $prefix)) {
+                    $prefix = substr($prefix, 0, -1);
+                }
+            }
+        }
+
+        if ($items !== []) {
+            $valuesBySharedPrefix[$prefix] = $items;
+            $items                         = [];
+            $prefix                        = null;
+        }
+
+        $regex = '(?>';
+
+        foreach ($valuesBySharedPrefix as $prefix => $items) {
+            if ($regex !== '(?>') {
+                $regex .= '|';
+            }
+
+            if (is_int($prefix)) {
+                $prefix = (string) $prefix;
+            }
+
+            $regex .= preg_quote($prefix, '/');
+
+            $regex .= count($items) === 1
+                ? preg_quote(substr(reset($items), strlen($prefix)), '/')
+                : $this->makeRegexFromList(array_map(static fn ($v) => substr($v, strlen($prefix)), $items), true);
+        }
+
+        return $regex . ')';
     }
 
     /**
      * Takes a SQL string and breaks it into tokens.
-     * Each token is an associative array with type and value.
      *
      * @param string $string The SQL string
      */
     public function tokenize(string $string): Cursor
     {
         $tokens = [];
 
-        $token = null;
+        $upper  = strtoupper($string);
```

Comment:
given that this optimizations in this method are begin kept, do you have a benchmark to understand the impact of all the other optimizations you are suggesting?

Reply:
Previously the uppercased matching was already done... and done very poorly as the string was upper cased again and again. The overall speedup of this PR is more than 90% - see https://github.com/doctrine/sql-formatter/actions/runs/9366932237/job/25785513399?pr=119#step:7:25 now and before. This is discussed in https://github.com/doctrine/sql-formatter/pull/119#issuecomment-2143747858 and the PR description.